src/GenOptProblem.hs

```diff
@@ -0,0 +1,706 @@
+-- | The GenOptProblem module performs several passes on the translation generated 
+-- by the Style compiler to generate the initial state (fields and GPIs) and optimization problem
+-- (objectives, constraints, and computations) specified by the Substance/Style pair.
+
+{-# OPTIONS_HADDOCK prune #-}
+{-# LANGUAGE AllowAmbiguousTypes, RankNTypes, UnicodeSyntax, NoMonomorphismRestriction #-}
+-- Mostly for autodiff
+
+module GenOptProblem where
+
+import Utils
+import Shapes
+import qualified Substance as C
+import Env
+import Style
+import Functions
+import Text.Show.Pretty (ppShow)
+import System.Random
+import Debug.Trace
+import qualified Data.Map.Strict as M
+import Control.Monad (foldM)
+import Data.List (intercalate, partition)
+import Data.Either (partitionEithers)
+
+-------------------- Type definitions
+
+type StyleOptFn = (String, [Expr]) -- Objective or constraint
+
+data OptType = Objfn | Constrfn
+     deriving (Show, Eq)
+
+data Fn = Fn { fname :: String,
+               fargs :: [Expr],
+               optType :: OptType }
+     deriving (Show, Eq)
+
+data FnDone a = FnDone { fname_d :: String,
+                         fargs_d :: [ArgVal a],
+                         optType_d :: OptType }
+     deriving (Show, Eq)
+
+-- A map from the varying path to its value, used to look up values in the translation
+type VaryMap a = M.Map Path (TagExpr a)
+
+------- State type definitions
+
+-- Stores the last EP varying state (that is, the state when the unconstrained opt last converged)
+type LastEPstate = [Float] -- Note: NOT polymorphic (due to system slowness with polymorphism)
+
+data OptStatus = NewIter
+               | UnconstrainedRunning LastEPstate
+               | UnconstrainedConverged LastEPstate
+               | EPConverged
+
+instance Show OptStatus where
+         show NewIter = "New iteration"
+         show (UnconstrainedRunning lastEPstate) =
+              "Unconstrained running with last EP state:\n" ++ show lastEPstate
+         show (UnconstrainedConverged lastEPstate) =
+              "Unconstrained converged with last EP state:\n" ++ show lastEPstate
+         show EPConverged = "EP converged"
+
+instance Eq OptStatus where
+         x == y = case (x, y) of
+                  (NewIter, NewIter) -> True
+                  (EPConverged, EPConverged) -> True
+                  (UnconstrainedRunning a, UnconstrainedRunning b) -> a == b
+                  (UnconstrainedConverged a, UnconstrainedConverged b) -> a == b
+                  (_, _) -> False
+
+data Params = Params { weight :: forall a . (Autofloat a) => a,
```

Comment:
Do weights need to be changed to `[Float]` as well? We are okay for now because we don't really use them (?)

Reply:
Done